### PR TITLE
Trigger replication failing

### LINKED ISSUE #2193: Trigger Replication Problem
Pasted form internal trello card

Neo4j Enterprise 4.2.9
APOC 4.2.0.2

Customer reported and I have since tested what seems to be a failure to replicate triggers created via apoc.trigger.add(). I’m aware of a bug in apoc that throws a server no longer accepts writes error unless you connect directly via bolt:// to the System db leader and add the trigger there (related cards and GH issue attached). But now, that trigger doesn’t appear anywhere else on the cluster, apart from that one core that I created it on. Is raft somehow being selective in what transactions its replicating across to members??  Since if I repeat the exercise with a simple create/merge node (also having connected directly via bolt to a user db) the merged node appears on all cores straightaway. I don’t think this is down to replication of updates in the system db, since other updates to system are replicated just fine. 

So the question is: What could possibly be preventing that trigger from appearing on nodes other than the one it was directly created on (system leader) using bolt://??

Interestingly, the trigger appears just fine on all cores, on a 3.5.29 cluster, which adds to my suspicion about how the System db updates in 4.x are being replicated, differently, to how graph.db updates were handled in 3.5.x

Attaching logs and relevant files from both test clusters (4.2.9 and 3.5.29). Please let me know if any additional details are required.

Repro steps:

1- connect via neo4j:// to a 4.x cluster (any 4.x version). Ensure beforehand to have the appropriate apoc plugin under /plugins directory and to whitelist apoc as follows in neo4j.conf:

```
apoc.trigger.enabled=true
dbms.security.procedures.whitelist=apoc.*
dbms.security.procedures.unrestricted=apoc.*
```

2- run `CALL apoc.trigger.add("trigger1", "MERGE (p:Person {name: 'invalid'})",{phase:'before'})`

unless one connects to instance that just happens to be the system db leader at the time, the following error will be thrown:

`Server at mydomainl:7687 no longer accepts writes`

This is a known bug with apoc, reported on the attached cards/GH.

3- CALL dbms.cluster.overview() and identify leader for the system db. Then connect via bolt:// directly (either via browser or cypher-shell) to that instance.

4- run `CALL apoc.trigger.add("trigger1", "MERGE (p:Person {name: 'invalid'})",{phase:'before'})`. This succeeds and one can view trigger1 as a listed trigger via call apoc.trigger.list().

5- Connect to any other core via bolt:// that is NOT the leader for system db and execute `call apoc.trigger.run`. Result: Trigger1 is NOT listed there, or on any other cluster members. This appears just fine on a 3.5.x cluster, but not in 4.x.

Thanks

### PR DESCRIPTION
**Pull request by conker84**
Fixes #2193

Fixes the replication problem 

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:

  - Fixed the bug
  - Improved docs
  - Added tests in cluster env

### CODE REVIEW

File: core/src/main/java/apoc/trigger/TriggerHandler.java [hunk 1]
```diff
@@ -138,26 +157,28 @@ private synchronized void reconcileKernelRegistration() {
...
+        withSystemDb(tx -> {
             tx.findNodes(SystemLabels.ApocTrigger,
-                    SystemPropertyKeys.database.name(), db.databaseName(),
-                    SystemPropertyKeys.name.name(), name)
+                            SystemPropertyKeys.database.name(), db.databaseName(),
+                            SystemPropertyKeys.name.name(), name)
                     .forEachRemaining(node -> node.setProperty(SystemPropertyKeys.paused.name(), paused));
-            tx.commit();
-        }
+            setLastUpdate(tx);
```
Comment (fbiville) [inline core/src/main/java/apoc/trigger/TriggerHandler.java:179]: why not call `setLastUpdate` in `withSystemDb` directly?

Comment (conker84) [reply core/src/main/java/apoc/trigger/TriggerHandler.java:179]: becuse we use `withSystemDb` also for reading ops like here:

https://github.com/neo4j-contrib/neo4j-apoc-procedures/blob/c1b34daf825933881ca91691dcb2752d3b914e20/core/src/main/java/apoc/trigger/TriggerHandler.java#L91-L105

**Comment by conker84** [comment]
@fbiville do another check I fixed a broken test

File: docs/asciidoc/modules/ROOT/partials/triggers.adoc [hunk 2]
```diff
@@ -5,4 +5,13 @@ We can enable them by setting the following property in `apoc.conf`:
...
\ No newline at end of file
+apoc.trigger.refresh=60000
+----
+
+.Description
+[opts=header,cols="3"]
+|===
+| Option Key | Value | Description
+| apoc.trigger.enabled | true/false, default false | Enable/Disable the feature
+| apoc.trigger.refresh | number, default 60000 | Value in ms for trigger replication check across cluster nodes
```
Comment (fbiville) [inline docs/asciidoc/modules/ROOT/partials/triggers.adoc:16]:
```suggestion
| apoc.trigger.refresh | number, default 60000 | Interval in ms after which a replication check is triggered across all cluster nodes
```

File: full/src/test/java/apoc/trigger/TriggerClusterTest.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    @Before
+    public void before() {
+        cluster.getSession().run("CALL apoc.trigger.removeAll()");
+        cluster.getSession().run("MATCH (n) DETACH DELETE n");
+    }
+
+
+    @Test
+    public void testTimeStampTriggerForUpdatedProperties() throws Exception {
+        cluster.getSession().run("CALL apoc.trigger.add('timestamp','UNWIND apoc.trigger.nodesByLabel($assignedNodeProperties,null) AS n SET n.ts = timestamp()',{})");
```
Comment (fbiville) [inline full/src/test/java/apoc/trigger/TriggerClusterTest.java:53]: wouldn't `$assignedLabels` make more sense here (and below)?

Comment (fbiville) [reply full/src/test/java/apoc/trigger/TriggerClusterTest.java:53]: Also, maybe for another PR, the short documentation of apoc.trigger.add still shows the old parameter syntax (https://neo4j.com/labs/apoc/4.0/overview/apoc.trigger/apoc.trigger.nodesByLabel/)

File: full/src/test/java/apoc/trigger/TriggerClusterTest.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        TestContainerUtil.testCall(cluster.getSession(), "MATCH (f:Foo) RETURN f", (row) -> {
+            assertEquals(true, ((Node)row.get("f")).containsKey("ts"));
+        });
+    }
+
+    @Test
+    public void testReplication() throws Exception {
+        cluster.getSession().run("CALL apoc.trigger.add('timestamp','UNWIND apoc.trigger.nodesByLabel($assignedNodeProperties,null) AS n SET n.ts = timestamp()',{})");
+
+        Thread.sleep(1000);
```
Comment (fbiville) [inline full/src/test/java/apoc/trigger/TriggerClusterTest.java:64]: do we have utilities such as `assertEventually` or something like that?

**Review by fbiville — APPROVED** [review]

File: full/src/test/java/apoc/trigger/TriggerClusterTest.java (hunk 4 above)
Comment (conker84) [reply full/src/test/java/apoc/trigger/TriggerClusterTest.java]: done

File: full/src/test/java/apoc/trigger/TriggerClusterTest.java (hunk 3 above)
Comment (conker84) [reply full/src/test/java/apoc/trigger/TriggerClusterTest.java:53]: oki

File: docs/asciidoc/modules/ROOT/partials/triggers.adoc (hunk 2 above)
Comment (conker84) [reply docs/asciidoc/modules/ROOT/partials/triggers.adoc]: done

**Review by fbiville — APPROVED** [review]

File: full/src/test/java/apoc/trigger/TriggerClusterTest.java (hunk 4 above)
Comment (fbiville) [reply full/src/test/java/apoc/trigger/TriggerClusterTest.java]: do you still need the call to Thread.sleep, then?

**Review by fbiville — APPROVED** [review]

File: full/src/test/java/apoc/trigger/TriggerClusterTest.java (hunk 4 above)
Comment (conker84) [reply full/src/test/java/apoc/trigger/TriggerClusterTest.java]: I think that you're looking at the old code because I removed it

**Pull request merged**